rp2040-hal/src/xosc.rs

```diff
@@ -0,0 +1,192 @@
+//! Crystal Oscillator (XOSC)
+// See [Chapter 2 Section 16](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
+
+use core::{
+    convert::Infallible,
+    ops::RangeInclusive
+};
+
+use embedded_time::{
+    fraction::Fraction,
+    fixed_point::FixedPoint,
+    rate::{
+        Hertz,
+        Megahertz,
+    },
+    duration::{
+        Seconds,
+        Milliseconds,
+        Duration
+    }
+};
+
+use nb::Error::WouldBlock;
+
+/// State of the Crystal Oscillator (typestate trait)
+pub trait State {}
+
+/// XOSC is disabled (typestate)
+pub struct Disabled;
+
+/// XOSC is initialized, ie we've given parameters (typestate)
+pub struct Initialized {
+    freq_hz: Hertz<u64>
+}
+
+/// Stable state (typestate)
+pub struct Stable{
+    freq_hz: Hertz<u64>
+}
+
+/// XOSC is disabling (typestate)
+pub struct Disabling;
+
+/// XOSC is in dormant mode (see Chapter 2, Section 16, §5)
+pub struct Dormant;
+
+impl State for Disabled {}
+impl State for Initialized {}
+impl State for Stable {}
+impl State for Disabling {}
+impl State for Dormant {}
+
+/// Possible errors when initializing the CrystalOscillator
+pub enum Error {
+    /// Frequency is out of the 1-15MHz range (see datasheet)
+    FrequencyOutOfRange,
+
+    /// Argument is bad : overflows, ...
+    BadArgument
+}
+
+/// A Crystal Oscillator.
+pub struct CrystalOscillator<S: State> {
+    device: rp2040_pac::XOSC,
+    state: S
+}
+
+impl<S: State> CrystalOscillator<S> {
+    /// Transitions the oscillator to another state.
+    fn transition<To: State>(self, state: To) -> CrystalOscillator<To> {
+        CrystalOscillator {
+            device: self.device,
+            state: state
+        }
+    }
+
+    /// Releases the underlying device.
+    pub fn free(self) -> rp2040_pac::XOSC {
+        self.device
+    }
+}
+
+impl CrystalOscillator<Disabled> {
+
+    /// Creates a new CrystalOscillator from the underlying device.
+    pub fn new(dev: rp2040_pac::XOSC) -> Self {
+        CrystalOscillator {
+            device: dev,
+            state: Disabled
+        }
+    }
+
+    /// Initializes the XOSC : frequency range is set, startup delay is calculated and set.
+    pub fn initialize(self, frequency: Megahertz<u32>) -> Result<CrystalOscillator<Initialized>, Error> {
+
+        const ALLOWED_FREQUENCY_RANGE: RangeInclusive<Megahertz<u32>> = Megahertz(1)..=Megahertz(15);
+        const STABLE_DELAY: Milliseconds<u64> = Milliseconds(1_u64);
+        const DIVIDER: Fraction = Fraction::new(1, 256);
+
+        if !ALLOWED_FREQUENCY_RANGE.contains(&frequency) {
+            return Err(Error::FrequencyOutOfRange)
+        }
+
+        self.device.ctrl.write(|w| {
+            w.freq_range()._1_15mhz();
+            w
+        });
+
+        let freq_hz: Hertz<u64> = frequency.into();
+        let delay_sec: Seconds<u64> = STABLE_DELAY.into();
+
+        //let startup_delay = ((freq_hz / 1000) + 128) / 256;
```

Comment:
A comment here is a good idea so the calculation is easier to understand, but I think the comment and the calculation do not match up. The divide ba 1000 seems to be missing below?

Reply:
Yes but no ! The calculation is actually a frequency in hertz being multiplied by a time in seconds. Except the time in second is 1ms, ie 10^-3 seconds, so it can be "simplified" to a divide by 1000 (and that's how it's written in the C SDK). Originally I was doing just the same but I felt it wasn't accurate regarding the formula so I brought back the multiply.
The comment is mathematically accurate but it doesn't reflect the calculation anymore so I'm going to change it :)